### ISSUE TITLE
Why do I get bad results when I remove prenet dropout in synthesis?

### ISSUE THREAD

**Issue by wugi88** [Rayhane-mamah/Tacotron-2#224]
Hello.
Thanks for sharing excellent work.

I want to always have same result when synthesizing.
So I tried to remove the dropout in the Prenet like below.

But, If I remove the dropout, the result of synthesis becomes so bad.
(I think it should get better result)

- tacotron/modules/modules.py
class Prenet:
	def __call__(self, inputs):

				# x = tf.layers.dropout(dense, rate=self.drop_rate, training=True,
				x = tf.layers.dropout(dense, rate=self.drop_rate, training=self.is_training,
					name='dropout_{}'.format(i + 1) + self.scope)
= When prenet dropout is exist.
![with_dropout_speech-alignment-00001](https://user-images.githubusercontent.com/29528818/46197101-c202f900-c343-11e8-9b4e-9dec81377f55.png)

= When prenet dropout is removed.
![without_dropout_speech-alignment-00001](https://user-images.githubusercontent.com/29528818/46197141-dfd05e00-c343-11e8-94ee-c853750762db.png)

I have no idea. Please help~

**Comment by candlewill** [comment]
I also tried to do this, but the result is good and the alignment is learnt well.

**Comment by xinqipony** [comment]
I also met the same problem, if dropout of prenet in inference time is removed, the voice quality and alignment are bad;

anyone knows why?
the model is not trained well?

as my understand, the dropout is used in training, even it's not used in inference, it should not affect quality ;

**Comment by begeekmyfriend** [comment]
When we talk about parameters we are talking about the variables of the whole NN. the weights the bias or something else. The framework will initialize all the variables by calling `global_variables_initializer`. And the variables would be initialized randomly rather than as a fixed value.

**Comment by xinqipony** [comment]
if we keep dropout in training, but remove dropout in inference, the alignment and quality becomes bad, this is what I am facing, @begeekmyfriend

**Comment by wugi88** [comment]
When I did the same thing on the NVIDA Tactron2, I had no problem.
Both sources have same 'Prenet' architecture.

Can anyone help me with what the prople is?

**Comment by Rayhane-mamah** [comment]
Hello @wugi88,
I faced similar problems once in the past and it then simply stopped happening before I could figure out potential reasons. Here's what I was suspecting:
- Overfit/Underfit: Location Sensitive Attention used by Tacotron-2 with previous alignments cumulation is highly sensitive to both sentences length (location aware) as well as sentences content (content aware). It also provides a top notch attention mechanism for close to production attention levels. It does however depend on the prenet quality a lot, which is introduced in the decoder, as an information bottleneck and a noise source for generalization during synthesis. If removing the prenet dropout ruins quality for you, it is a good sign that your prenet is facing some potential overfitting. With that in mind could you please provide your current training steps, and possibly the tensorboard evaluation curves? Normally L2 regularization should help with this.
- Explosions: did by any chance, any gradient explosion occur during your run?
- Normally, making the change you made should take care of required scalings etc, but could you also instead of modified the "training" parameter, change instead the drop_rate to 0?

Like I said earlier, similar problem no longer occurs in my tests for some reason, so any feedback on your side can help pinpoint the cause very much! Thanks for your contribution @wugi88 !

**Comment by wugi88** [comment]
Hello~ @Rayhane-mamah
Thank you for you kind reply.

Even if I set the 'drop_rate' to 0, I got a same synthesis results as 'training=False'.
Here are tensorboard charts.

![2018-10-01 16-37-25](https://user-images.githubusercontent.com/29528818/46275747-f58a9100-c598-11e8-88f9-5acbea5f240e.png)

![2018-10-01 16-38-02](https://user-images.githubusercontent.com/29528818/46275709-d7249580-c598-11e8-9883-28caf01ddd35.png)

But If I scale the 'Prenet' like below. synthesis makes good results.
As you told me, if it doesn't happen anymore. Could it be tensorflow dropout bug?

				x = tf.layers.dropout(dense, rate=self.drop_rate, training=self.is_training,
									  name='dropout_{}'.format(i + 1) + self.scope)

				x = x/(1-self.drop_rate)

= When above code is applied.
![without_dropout_divided_alignment-1](https://user-images.githubusercontent.com/29528818/46276468-253a9880-c59b-11e8-9cfe-6510b217d0c0.png)

**Comment by xinqipony** [comment]
@wugi88  in your trying:
x = tf.layers.dropout(dense, rate=self.drop_rate, training=self.is_training,name='dropout_{}'.format(i + 1)+ self.scope)
x = x/(1-self.drop_rate)

you keep the drop_ratio as 0.5 for line 1 and line 2?

**Comment by wugi88** [comment]
@xinqipony Yes, it is.

**Comment by wugi88** [comment]
I think my idea was totally wrong. T.T

As Rayhane-mamah said,
I am experimenting with two cases.

1) add L2 regularization to the 'Prenet'.
2) change tacotron_reg_weight to 1e-5

When I get the results, I'll update them.

**Comment by Rayhane-mamah** [comment]
Hey,

Normally, L2 reg is automatically applied to the prenet, on LJspeech 1e-7
seemed reasonable.
It's possible that your problem is directly related to the number of
training steps so here is what I suggest.
Change "True" to "is_training" so that during eval, entire prenet is used.
Keep monitoring attention during eval steps and ensure it does not overfit.
If it starts to overfit simply stop training.

Let me know if eval attention never gets aligned (with full prenet)

**Comment by wugi88** [comment]
Hello~ @Rayhane-mamah 

By the way, I'm training your model with different dataset (a Korean dataset).
Your work makes very good quality speeches : [here is a synthesis sample](https://wugi88.github.io/Tacotron-WaveNet-samples/samples/son_fortune.wav)

Because of the prenet dropout results, I thought I should make the Prenet simpler.
So I try to add extra l2 reg as below.

            # Get all trainable variables
            all_vars = tf.trainable_variables()
            regularization = tf.add_n([tf.nn.l2_loss(v) for v in all_vars
                if not('bias' in v.name or 'Bias' in v.name)]) * reg_weight

            # Get Decoder Prenet variables
            prenet_vars = tf.trainable_variables(scope = 'model/inference/decoder/decoder_prenet')
            prenet_regularization = tf.add_n([tf.nn.l2_loss(v) for v in prenet_vars
                if not('bias' in v.name or 'Bias' in v.name)]) * hp.tacotron_prenet_reg_weight
           ......
            self.regularization_loss = regularization + prenet_regularization

And here are some results.. (and other combinations are also being tried.)

1) all reg 1e-5, and extra prenet reg 0
![ray3 prenet_0 reg_1e-5](https://user-images.githubusercontent.com/29528818/46591334-02f1cf00-caf5-11e8-9e2b-10559d027f00.png)

2) all reg 1e-6, and extra prenet reg 1e-2
![ray1 prenet_1e-2 reg_1e-6](https://user-images.githubusercontent.com/29528818/46591346-1c931680-caf5-11e8-9975-fce2758bb259.png)

3) all reg 1e-6, and extra prenet reg 1e-1
![ray2 prenet_1e-1 reg_1e-6](https://user-images.githubusercontent.com/29528818/46591364-33396d80-caf5-11e8-8afb-cc827cd7cfc6.png)

When I get the results, I'll update them.

**Comment by xinqipony** [comment]
@Rayhane-mamah , is the prenet issue here also happening in your model? for your latest code, do we have a model to test maybe?

**Comment by xinqipony** [comment]
@wugi88 , any update?

**Comment by wugi88** [comment]
Hello @xinqipony .
Here are some experimental results.

![2018-10-15 12 55 01](https://user-images.githubusercontent.com/29528818/46929417-858d0800-d07a-11e8-989b-0eef364f1f8c.png)

On my dataset 'tacotron_reg_weight=1e-4' works well (the green one).

**Comment by xinqipony** [comment]
@wugi88 do  you use extra prenet reg?

**Comment by wugi88** [comment]
@xinqipony Sorry. I didn't check the synthesis result.
It's not working.

**Comment by xinqipony** [comment]
@Rayhane-mamah any insight?

**Comment by xinqipony** [comment]
@wugi88  just let you know, per my test result, I tried using tanh in prenet, yes, the dropout in prenet can be removed now, but the overall quality is not good as using relu and dropout, so this may still need more optimization, I m still trying

**Comment by wugi88** [comment]
@xinqipony Try early stop. (To avoid overfitting, or get more data)

**Comment by XiaooaiX** [comment]
Sadly I also met the same problem these days T.T..  Have read through this issue, however I don't think it's the problem of activation function or overfitting because in other repos(such as those implemented by NVIDA) this problem doesn't exist, there must be some other problems that we have not found. 
Anyone has some ideas?

**Comment by candlewill** [comment]
Here is my solution, very tricky:

```python
drop_rate = tf.random_uniform(shape=[1], minval=0, maxval=min(self.drop_rate * 1.5, 0.5), dtype=tf.float32)[0]
x = tf.layers.dropout(dense, rate=drop_rate, training=self.is_training,
				                name='dropout_{}'.format(i + 1) + self.scope)
```

**Comment by pravn** [comment]
Prenet is a way to regularize tacotron. Since the training uses teacher forcing to use ground truth samples, the model needs some other mechanism to generalize when the samples are not from the training set. I believe the paper mentions scheduled sampling (using the previous decoder output with a schedule) and argues that the results are poor. So it's important to regularize with dropout in order for the model to perform well during inference mode. On the same lines, we might expect that a smaller dataset might need more dropout included to penalize it appropriately.

**Comment by begeekmyfriend** [comment]
One more argument. In my humble opinion, the dropouts in [convolution](https://github.com/Rayhane-mamah/Tacotron-2/blob/master/tacotron/models/modules.py#L390) can be remove according to such articles like this [blog](https://towardsdatascience.com/dont-use-dropout-in-convolutional-networks-81486c823c16) and this [reddit post](https://www.reddit.com/r/MachineLearning/comments/42nnpe/why_do_i_never_see_dropout_applied_in/). In convolution layers there are less parameters and the regularization that the dropout layer bring might not work well. Furthermore the convolution layers can be found in encoder, postnet and CBHG which might have little effect on mel spectrograms prediction. In contrast the dropout layer in decoder prenet and zoneout LSTM can be kept as I suggest. The loss obviously drop when I do it. Let us wait and see the final results.
![step-1000-align](https://user-images.githubusercontent.com/6031938/55671460-c4ca4200-58c2-11e9-946d-2b5e79c1ddbc.png)

**Comment by mrgloom** [comment]
I wonder what is the proper way to get reproducible results at inference time?

Here is dropout that is used at inference time also (`training=True`)
https://github.com/Rayhane-mamah/Tacotron-2/blob/4dc16ce1dd08d81534bc4b218c8cd450183c3f3c/tacotron/models/modules.py#L249

Setting `tf.set_random_seed(2019)` in main script that I run don't helped or setting at top of https://github.com/Rayhane-mamah/Tacotron-2/blob/master/tacotron/models/modules.py , but setting random seed inside function helped:

```
    def __call__(self, inputs):
        x = inputs

        # Fix random seed for reprodicibility
        tf.set_random_seed(2019)

        with tf.variable_scope(self.scope):
            for i, size in enumerate(self.layers_sizes):
                dense = tf.layers.dense(x, units=size, activation=self.activation,
                                        name="dense_{}".format(i + 1))
                # The paper discussed introducing diversity in generation at inference time
                # by using a dropout of 0.5 only in prenet layers (in both training and inference).
                x = tf.layers.dropout(dense, rate=self.drop_rate, training=True,
                                      name="dropout_{}".format(i + 1) + self.scope)
        return x
```

**Comment by mrgloom** [comment]
Here is simplified example of what I mean by fixing random seed in main script:

run_me.py

```
from file1 import f1

import tensorflow as tf
tf.random.set_random_seed(2019)

print('tf.__version__', tf.__version__)

if __name__ == "__main__":
    f1()
```

file1.py
```
import numpy as np

import os
os.environ['TF_CPP_MIN_LOG_LEVEL']='1'

import tensorflow as tf
tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)

np.random.seed(2019)

def f1():
    x = tf.placeholder(tf.float32, shape=(1,2))
    y = tf.layers.dense(x, units=6, activation=None, name='dense_0')
    y = tf.layers.dropout(y, rate=0.5, training=True, name='dropout_0')

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        x_np = np.random.rand(1,2)
        print(sess.run(y, feed_dict={x: x_np}))
        print(sess.run(y, feed_dict={x: x_np}))
```

`tf.__version__ 1.14.0`

run1:
```
[[ 0.          0.          0.76269484  0.26875716 -0.94779396 -0.30788693]]
[[ 0.3769264   1.3462708   0.          0.         -0.94779396 -0.        ]]
```

run2:
```
[[ 0.          0.          0.76269484  0.26875716 -0.94779396 -0.30788693]]
[[ 0.3769264   1.3462708   0.          0.         -0.94779396 -0.        ]]
```
